File: packages/python/corti_agent_sdk/response.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from .types import Artifact, Message, Task


class MessageResponse:
    """
    Wraps the raw ``message:send`` API response and promotes the fields you
    almost always need to the top level, while keeping the full response
    accessible via ``.raw``.

    Example::

        r = await ctx.send_text("What is the ICD-10 code for hypertension?")
        print(r.text)      # "The ICD-10 code is I10."
        print(r.status)    # "completed"
        print(r.task)      # full A2A v1 Task dict
        print(r.raw)       # full API response dict
    """

    def __init__(self, raw: Dict[str, Any]) -> None:
        self._raw = raw

    @classmethod
    def from_text(cls, text: str) -> "MessageResponse":
        """
        Synthesise a completed ``MessageResponse`` from a plain string.
        Used internally when merging parallel results into a single response.
        """
        return cls({
            "task": {
                "id": "",
                "contextId": "",
                "kind": "task",
                "status": {
                    "state": "completed",
                    "message": {
                        "role": "agent",
                        "parts": [{"kind": "text", "text": text}],
                        "messageId": "",
                        "kind": "message",
                    },
                },
            }
        })

    # ── private helpers ───────────────────────────────────────────────────────

    @property
    def _node(self) -> Dict[str, Any]:
        return self._raw.get("task") or {}

    @property
    def _node_status(self) -> Dict[str, Any]:
        return self._node.get("status") or {}

    # ── A2A v1 task ───────────────────────────────────────────────────────────

    @property
    def task(self) -> Optional[Task]:
        """The full A2A v1 Task object."""
        return self._raw.get("task")  # type: ignore[return-value]

    # ── promoted fields ───────────────────────────────────────────────────────

    @property
    def status(self) -> Optional[str]:
        """The task's terminal state, e.g. ``"completed"``, ``"failed"``."""
        return self._node_status.get("state")

    @property
    def status_message(self) -> Optional[Message]:
        """The agent's reply message (``task.status.message``)."""
        return self._node_status.get("message")  # type: ignore[return-value]

    @property
    def text(self) -> Optional[str]:
        """All text parts from ``status_message`` joined into a single string."""
        msg = self.status_message
        if not msg:
            return None
        parts = msg.get("parts") or []  # type: ignore[union-attr]
        joined = "".join(
            p["text"] for p in parts if p.get("kind") == "text" and "text" in p
        )
        return joined or None

    @property
    def artifacts(self) -> List[Artifact]:
        """Structured artifacts produced by the task, deduplicated by parts content."""
        seen: set[str] = set()
        result: List[Artifact] = []
        for a in self._node.get("artifacts") or []:
            key = json.dumps(a.get("parts"), sort_keys=True)
            if key not in seen:
                seen.add(key)
                result.append(a)
        return result

    @property
    def context_id(self) -> Optional[str]:
        """The thread ID — same value the context tracks internally."""
        return self._node.get("contextId")

    @property
    def task_id(self) -> Optional[str]:
        """The task ID for this specific invocation."""
        return self._node.get("id")

    @property
    def raw(self) -> Dict[str, Any]:
        """The full, unmodified response from the API."""
        return self._raw

File: packages/js/src/response.ts
import type { Corti } from "@corti/sdk";
import type { SendMessageResponse } from "./types.js";

type Task = Corti.CommonTaskResponse;
type Message = Corti.CommonMessage;
type Artifact = Corti.CommonArtifactResponse;
type TaskState = Corti.CommonTaskState;
type Part = Corti.CommonPart;

export class MessageResponse {
  private readonly _task: Task | undefined;
  private readonly _message: Message | undefined;

  constructor(response: SendMessageResponse) {
    this._task = response.task;
    this._message = response.message;
  }

  static fromText(text: string): MessageResponse {
    return new MessageResponse({
      message: {
        role: "ROLE_AGENT",
        parts: [{ text }],
      },
    });
  }

  static fromTask(task: Task): MessageResponse {
    return new MessageResponse({ task });
  }

  get task(): Task | undefined {
    return this._task;
  }

  get message(): Message | undefined {
    if (this._message) return this._message;
    return this._task?.status?.message;
  }

  get taskId(): string | undefined {
    return this._task?.id;
  }

  get contextId(): string | undefined {
    return this._task?.contextId;
  }

  get state(): TaskState | undefined {
    return this._task?.status?.state;
  }

  get status(): "completed" | "failed" | "working" | "submitted" | "canceled" | "input-required" | "rejected" | "auth-required" | undefined {
    const state = this.state;
    if (!state) return undefined;
    return state.replace("TASK_STATE_", "").toLowerCase().replace(/_/g, "-") as
      | "completed"
      | "failed"
      | "working"
      | "submitted"
      | "canceled"
      | "input-required"
      | "rejected"
      | "auth-required";
  }

  get text(): string | null {
    const msg = this.message;
    if (!msg || msg.role === "ROLE_USER") return null;
    const joined = (msg.parts ?? [])
      .filter((p: Part): p is { text: string } => "text" in p && typeof p.text === "string")
      .map((p) => p.text)
      .join("");
    return joined === "" ? null : joined;
  }

  get artifacts(): Artifact[] {
    const all = this._task?.artifacts ?? [];
    const seen = new Set<string>();
    return all.filter((a) => {
      const key = JSON.stringify(a.parts);
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }

  get raw(): SendMessageResponse {
    if (this._task) return { task: this._task };
    if (this._message) return { message: this._message };
    return {};
  }
}
